### DISCUSSION TITLE
How to to avoid `Y@Foo cannot be assigned to Y@Foo | <something>`?

### DISCUSSION THREAD

**Discussion by patrick-kidger** [microsoft/pyright#5599] [Q&A]
I can't seem to find a way to pass this generic class instance to an annotation of the same generic class.

This:
```python
from collections.abc import Callable
from typing import Any, Generic, TypeVar

Y = TypeVar("Y")

class Foo(Generic[Y]):
    def __init__(self, operation: Callable[[Y], Y]):
        pass

def call(foo: Foo[Y], y: Y) -> Y:
    ...

foo = Foo(lambda x: x)
call(foo, 1.)
call(foo, "hi")
```
produces:
```
file.py
  file.py:14:6 - error: Argument of type "Foo[Y@Foo]" cannot be assigned to parameter "foo" of type "Foo[Y@call]" in function "call"
    "Foo[Y@Foo]" is incompatible with "Foo[Y@Foo | float]"
      TypeVar "Y@Foo" is invariant
        Type "Y@Foo" cannot be assigned to type "Y@Foo | float" (reportGeneralTypeIssues)
  file.py:15:6 - error: Argument of type "Foo[Y@Foo]" cannot be assigned to parameter "foo" of type "Foo[Y@call]" in function "call"
    "Foo[Y@Foo]" is incompatible with "Foo[Y@Foo | str]"
      TypeVar "Y@Foo" is invariant
        Type "Y@Foo" cannot be assigned to type "Y@Foo | str" (reportGeneralTypeIssues)
2 errors, 0 warnings, 0 informations 
```

What's the appropriate fix for this?

I've included `call` against two different types to emphasise that `foo = Foo[float](lambda x: x)` won't fix this. I'm looking for behaviour just like callables, where a single generic instance can be used with different concrete types (`foo = lambda x: x; foo(1); foo("hi")` is valid).

The reason is that `Foo` and `call` are provided by my library, for which I want to do the Right Thing and use pyright throughout. But the last few lines, instantiating `foo` and calling `call`, will be done by consumers of my library -- who are typically data scientists that aren't terribly interested in the finer details of Python's type system! I'd like things to be usable for them.

**Comment by erictraut** [comment]
When you construct a `Foo`, the constructed type must be concrete — i.e. the type argument `Y` needs to have a concrete type associated with it. When you pass it a `lambda x: x` as an argument, there's no way for a static type checker to "solve for `Y`", so it goes unsolved. You would need to explicitly specialize it if you want to use this technique.

```python
foo = Foo[float](lambda x: x)
call(foo, 1.0)
call(foo, "") # Type error
```

Or you could pass it a callable that has defined parameter and return types:

```python
def identity(x: float) -> float:
    return x

foo = Foo(identity)
call(foo, 1.0)
call(foo, "") # Type error
```

If you want `Foo` to encapsulate a callable that accepts any type and returns the same type, then you shouldn't use a class-scoped type parameter.

```python
from typing import Any, Callable, TypeVar

Y = TypeVar("Y")

class Foo:
    def __init__(self, operation: Callable[[Y], Y]):
        self._operaton: Callable[[Any], Any] = operation

    def __call__(self, x: Y) -> Y:
        return self._operaton(x)

def call(foo: Foo, y: Y) -> Y:
    return foo(y)

foo = Foo(lambda x: x)

v1 = call(foo, 1.0)
v2 = call(foo, "hi")
```

**Reply by patrick-kidger** [reply]
Ah, so functions are special-cased as "generic instances"? That is,
```python
def foo(x: Y) -> Y:
    ...
```
is an instance of `types.FunctionType` with type parameters that may be resolved at call sites, rather than at construction time.

My assumption was that this resolution-at-call-sites was true for all types -- I didn't realise there was a notion of unresolved types.

Anyway, thankyou for the help as always. I'll take a closer look at my application and see if I can resolve it in the way you suggest.
